pkg/controller/volume/attach_detach_controller.go

```diff
@@ -106,30 +240,365 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+	}
+
+	// Process "safe to detach" volumes annotations. Mark volumes in "actual
+	// state of world" as safe to detach. Delete annotations for the rest.
+	adc.processSafeToDetachAnnotations(nodeName, node.Annotations)
 }
 
 func (adc *attachDetachController) nodeUpdate(oldObj, newObj interface{}) {
-	// No op for now
+	// The flow for update is the same as add.
+	adc.nodeAdd(newObj)
 }
 
 func (adc *attachDetachController) nodeDelete(obj interface{}) {
-	// No op for now
+	node, ok := obj.(*api.Node)
+	if node == nil || !ok {
+		return
+	}
+
+	nodeName := node.Name
+	if err := adc.desiredStateOfWorld.DeleteNode(nodeName); err != nil {
+		glog.V(10).Infof("%v", err)
+	}
+
+	// Process "safe to detach" volumes annotations. Mark volumes in "actual
+	// state of world" as safe to detach. Delete annotations for the rest.
+	adc.processSafeToDetachAnnotations(nodeName, node.Annotations)
+}
+
+// processPodVolumes processes the volumes in the given pod and adds or deletes
+// from the desired state of the world as appropriate
+func (adc *attachDetachController) processPodVolumes(
+	pod *api.Pod, ignorePhaseDeleteVolumes bool) {
+	if pod == nil {
+		return
+	}
+
+	if len(pod.Spec.Volumes) <= 0 {
+		return
+	}
+
+	if !adc.desiredStateOfWorld.NodeExists(pod.Spec.NodeName) {
+		// If the node the pod is scheduled to does not exist in the desired
+		// state of the world data structure, that indicates the node is not
+		// yet managed by the controller. Therefore, ignore the pod.
+		// If the node is added to the list of managed nodes in the future,
+		// future adds and updates to the pod will be processed.
+		glog.V(10).Infof(
+			"Skipping processing of pod %q/%q: it is scheduled to node %q which is not managed by the controller.",
+			pod.Namespace,
+			pod.Name,
+			pod.Spec.NodeName)
+		return
+	}
+
+	// Process volume spec for each volume defined in pod
+	for _, podVolume := range pod.Spec.Volumes {
+		volumeSpec, err := adc.createVolumeSpec(podVolume, pod.Namespace)
+		if err != nil {
+			glog.V(10).Infof(
+				"Error processing volume %q for pod %q/%q: %v",
+				podVolume.Name,
+				pod.Namespace,
+				pod.Name,
+				err)
+			continue
+		}
+
+		attachableVolumePlugin, err := adc.isAttachableVolumePlugin(volumeSpec)
+		if err != nil || attachableVolumePlugin == nil {
+			glog.V(10).Infof(
+				"Skipping volume %q for pod %q/%q: it does not implement attacher interface.",
+				podVolume.Name,
+				pod.Namespace,
+				pod.Name)
+			continue
+		}
+
+		if !ignorePhaseDeleteVolumes &&
+			(pod.Status.Phase == api.PodPending ||
+				pod.Status.Phase == api.PodRunning) {
+			// Add volume to desired state of world
+			_, err := adc.desiredStateOfWorld.AddPod(
+				getUniquePodName(pod), volumeSpec, pod.Spec.NodeName)
+			if err != nil {
+				glog.V(10).Infof(
+					"Failed to add volume %q for pod %q/%q to desiredStateOfWorld. %v",
+					podVolume.Name,
+					pod.Namespace,
+					pod.Name,
+					err)
+			}
+
+			continue
+		}
+
```

Comment:
It looks like this is the else of the above.. I think it would be clearer if there an explicit check on the phases handled here. Maybe even throw an error for an unknown phase.